haskell/private/cc_wrapper.py.tpl

```diff
@@ -0,0 +1,812 @@
+                self._handle_rpath(ld_args[1], out)
+                return True
+            else:
+                out.append(arg)
+                return True
+        else:
+            return False
+
+    def _handle_rpath(self, rpath, out):
+        # Filter out all RPATH flags for now and manually add the needed ones
+        # later on.
+        self.rpaths.append(rpath)
+
+    def _handle_print_file_name(self, arg, args, out):
+        if arg == "--print-file-name":
+            print_file_name = next(args)
+        elif arg.startswith("--print-file-name="):
+            print_file_name = arg[len("--print-file-name="):]
+        else:
+            return False
+
+        # Remember print-file-name action. Don't forward to allow for later
+        # manipulation.
+        self.print_file_name = print_file_name
+        self.action = Args.PRINT_FILE_NAME
+
+        return True
+
+    def _handle_compile(self, arg, args, out):
+        if arg == "-c":
+            self.action = Args.COMPILE
+            out.append(arg)
+        else:
+            return False
+
+        return True
+
+
+def load_response_files(args):
+    """Generator that loads arguments from response files.
+
+    Passes through any regular arguments.
+
+    Args:
+      args: Iterable of arguments.
+
+    Yields:
+      All arguments, with response files replaced by their contained arguments.
+
+    """
+    args = iter(args)
+    for arg in args:
+        if arg == "-install_name":
+            # macOS only: The install_name may start with an '@' character.
+            yield arg
+            yield next(args)
+        elif arg.startswith("@"):
+            with open(arg[1:], "r") as rsp:
+                for line in rsp:
+                    for rsp_arg in parse_response_line(line):
+                        yield rsp_arg
+        else:
+            yield arg
+
+
+def parse_response_line(s):
+    # GHC writes response files with quoted lines.
+    return shlex.split(s)
+
+
+def shorten_path(input_path):
+    """Shorten the given path if possible.
+
+    Applies the following transformations if they shorten the path length:
+      - Make path relative to CWD.
+      - Remove redundant up-level references.
+      - Resolve symbolic links.
+
+    Args:
+      input_path: The path to shorten.
+
+    Returns:
+      The shortened path.
+
+    """
+    exists = os.path.exists(input_path)
+    shortened = input_path
+
+    # Try relativizing to current working directory.
+    rel = os.path.relpath(shortened)
+    if len(rel) < len(shortened):
+        shortened = rel
+
+    # Try normalizing the path if possible.
+    norm = os.path.normpath(shortened)
+    if len(norm) < len(shortened):
+        # Ensure that the path is still correct. Reducing up-level references
+        # may change the meaning of the path in the presence of symbolic links.
+        try:
+            if not exists or os.path.samefile(norm, shortened):
+                shortened = norm
```

Comment:
This is where allowing non-existent files becomes questionable, because what if we screw up the non-existent path here and don’t notice. Could lead to nasty bugs.

Reply:
See reply in https://github.com/tweag/rules_haskell/pull/1039/files/71c2a6d5286a0e0802c78be3a1a8cf6e6ae8061d..a10b82374bc7b6f86844835b1656e06b449a0b0a#r318499294